src/Microsoft.AspNetCore.Mvc.Api.Analyzers/SymbolApiResponseMetadataProvider.cs

```diff
@@ -276,16 +276,150 @@ internal static int GetStatusCode(AttributeData attribute)
+                // Covers the 'return StatusCode(200)' case.
+                if (TryGetParameterStatusCode(semanticModel, invocation.Expression, invocation.ArgumentList, cancellationToken, out var statusCode))
+                {
+                    return new ActualApiResponseMetadata(returnStatementSyntax, statusCode);
+                }
+
+                return null;
+            }
+
+            if (returnExpression is ObjectCreationExpressionSyntax creation)
+            {
+                // Covers the 'return new ObjectResult(...) { StatusCode = 200 }' case.
+                if (TryGetInitializerStatusCode(semanticModel, creation.Initializer, cancellationToken, out var statusCode))
+                {
+                    return new ActualApiResponseMetadata(returnStatementSyntax, statusCode);
+                }
+
+                // Covers the 'return new StatusCodeResult(200) case.
+                if (TryGetParameterStatusCode(semanticModel, creation, creation.ArgumentList, cancellationToken, out statusCode))
+                {
+                    return new ActualApiResponseMetadata(returnStatementSyntax, statusCode);
+                }
+            }
+
             return null;
         }
 
+        private static bool TryGetInitializerStatusCode(
+            SemanticModel semanticModel,
+            InitializerExpressionSyntax initializer,
+            CancellationToken cancellationToken,
+            out int statusCode)
+        {
+            if (initializer == null)
+            {
+                statusCode = default;
+                return false;
+            }
+
+            foreach (var assignment in initializer.Expressions.OfType<AssignmentExpressionSyntax>())
+            {
+                if (assignment.Left is IdentifierNameSyntax identifier)
+                {
+                    var symbolInfo = semanticModel.GetSymbolInfo(identifier, cancellationToken);
+
+                    if (symbolInfo.Symbol is IPropertySymbol property && property.Name.Equals("StatusCode"))
+                    {
+                        return TryGetExpressionStatusCode(semanticModel, assignment.Right, cancellationToken, out statusCode);
+                    }
+                }
+            }
+
+            statusCode = default;
+            return false;
+        }
+
+        private static bool TryGetParameterStatusCode(
+            SemanticModel semanticModel,
+            ExpressionSyntax expression,
+            BaseArgumentListSyntax argumentList,
+            CancellationToken cancellationToken,
+            out int statusCode)
+        {
+            var symbolInfo = semanticModel.GetSymbolInfo(expression, cancellationToken);
+
+            if (symbolInfo.Symbol is IMethodSymbol method)
+            {
+                var statusCodeParameter = method.Parameters.FirstOrDefault(parameter =>
+                    parameter.Type.SpecialType == SpecialType.System_Int32 &&
+                    parameter.Name.Equals("statusCode"));
```

Comment:
This simple heuristic is more or less duplicated several times. Is looking for a property/parameter named `StatusCode`/`statusCode` of type `Int32` enough, or should we introduce an additional attribute, like the `DefaultStatusCodeAttribute`?

The current behavior has the added benefit that overloads like

https://github.com/aspnet/Mvc/blob/da1189e6f12c485ddda9c5cefe01b8cde2708f3c/src/Microsoft.AspNetCore.Mvc.Core/ControllerBase.cs#L213-L219

would Just Work™

Reply:
> or should we introduce an additional attribute, like the DefaultStatusCodeAttribute?

That would certainly reduce the risk of false positives (e.g. `var orderStatusDescription = GetStatusAsync(status: order.StatusCode);`). We could start with an `internal` attribute decorated on the parameter \ property.

@glennc \ @rynowak thoughts on this?

Reply:
Just to be clear, that specific case wouldn't be flagged as a false positive.

The rules are:

- It must be a return statement
- The returned type must be assignable to `IActionResult`
- The name of the parameter must be `statusCode` (or `StatusCode` in the object initializer case)
- The type must be `Int32`

I think the chances of false positives are slim, but having an attribute would remove that risk.

Reply:
> The returned type must be assignable to IActionResult

Ahh, missed that bit. Yeah, that does generally remove the risk, but I guess we could always start being precise \ conservative and expand the analyzer if it misses returned values pretty frequently.